### PR TITLE
feat: improve storage

### PR DESCRIPTION
**Pull request by pi0**
[universal-storage](https://github.com/nuxt-community/universal-storage) is using a forked and enhanced version of auth-module's storage. Main improvements:

- Universal cookies support
- Auto object and primitives serialization and deserialization
- Heavily tested on alibaba.aero

### CODE REVIEW

File: lib/core/storage.js [hunk 1]
```diff
@@ -128,79 +132,101 @@ export default class Storage {
...
-      return false
-    }
-    return isJson ? JSON.parse(value) : value
+    const value = cookies[_key] ? decodeURIComponent(cookies[_key]) : undefined
+
+    return decodeValue(value)
+  }
+
+  removeCookie (key, options) {
+    this.setCookie(key, undefined, Object.assign({ maxAge: -1 }, options))
```
Comment (MathiasCiarlo) [inline lib/core/storage.js]: What about changes in the rest of the auth module? For instance - the new removeUniversal helper should now be called from [this line in auth.js](https://github.com/nuxt-community/auth-module/blob/dev/lib/core/auth.js#L350). If not - the cookie will not be removed, but simply set to "null", without expiration date.

File: lib/core/storage.js [hunk 2]
```diff
@@ -128,79 +132,101 @@ export default class Storage {
...
     const _key = this.options.cookie.prefix + key
-
     const _options = Object.assign({}, this.options.cookie.options, options)
+    const _value = encodeValue(value)
 
-    if (isUnset(value)) {
-      Cookies.remove(_key, _options)
-    } else {
-      Cookies.set(_key, value, _options)
+    const serializedCookie = serializeCookie(_key, _value, _options)
```
Comment (MathiasCiarlo) [inline lib/core/storage.js]: Are you sure the js-cookie options will work with cookie? @sambrezo said the options (e.g. expires) are used differently in the two libraries. See his comment [here](https://github.com/nuxt-community/auth-module/pull/333#issuecomment-487433164)

File: lib/core/utilities.js [hunk 3]
```diff
@@ -63,3 +63,37 @@ export function normalizePath (path = '') {
...
+export function encodeValue (val) {
+  if (typeof val === 'string') {
+    return val
+  }
+  return JSON.stringify(val)
+}
+
+export function decodeValue (val) {
+  // Find current value's type
+  const type = Object.prototype.toString.call(val)
```
[the remark this answers is not in the document]
Comment (pi0) [reply lib/core/utilities.js]: Seems reasonable. (I hope we are not missing any point as this change was almost 1 year ago :D)

[the remark this answers is not in the document]
Comment (pi0) [reply lib/core/utilities.js]: /cc @farzadso if you can remember reason

File: lib/core/storage.js (hunk 1 above)
Comment (pi0) [reply lib/core/storage.js]: Will update other areas as well. Thanks for the tip. (however, I think we should bump to 5.x after this PR as this is a breaking change)

File: lib/core/storage.js (hunk 2 above)
Comment (pi0) [reply lib/core/storage.js]: I'm not sure to get the point. We are now using [cookie](https://www.npmjs.com/package/cookie) not [js-cookie](https://www.npmjs.com/package/js-cookie). `cookie` package allows setting options like `expires`

Comment (MathiasCiarlo) [reply lib/core/storage.js]: Ah, my bad. Didn't catch that you removed the js-cookie lib :)

File: lib/core/utilities.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (farzadso) [reply lib/core/utilities.js]: To be honest, I can't remember the exact reason. I can remember the other comments but not this exact line.

I think we were trying to differentiate between Arrays and Objects but it seems like we didn't finish it.

File: lib/core/storage.js (hunk 1 above)
Comment (pi0) [reply lib/core/storage.js]: (for now, added backward compatible conditions to not be a breaking change)

**Comment by pi0** [comment]
This PR has been published in [v4.6.0](https://github.com/nuxt-community/auth-module/releases/tag/v4.6.0)

File: lib/core/storage.js (hunk 2 above)
Comment (enVolt) [reply lib/core/storage.js]: Wait, this breaks the functionality where cookie expires can be defined as number in days. cookie expects fixed expiry time. js-cookie if provided number convert that to `Date.now() + n Days` (https://github.com/js-cookie/js-cookie/blob/master/src/js.cookie.js#L55)

With new release this functionality is broken (v4.6.0)

This should have been a major version bump.

Comment (pi0) [reply lib/core/storage.js]: @enVolt Thanks for this point. I'll add number support for setCookie options to prevent breaking changes.

Comment (pi0) [reply lib/core/storage.js]: https://github.com/nuxt-community/auth-module/releases/tag/v4.6.1

**Pull request merged**